ekncontent/ekncontent/eknc-domain.c

```diff
@@ -739,36 +739,134 @@ static void
 query_state_free (gpointer data)
 {
   QueryState *state = data;
-  g_slist_free_full (state->models, g_object_unref);
+  g_clear_pointer (&state->models, g_ptr_array_unref);
   g_slice_free (QueryState, state);
 }
 
+typedef void (*ScatterTaskCountdownCompleteCallback)(gpointer user_data);
+typedef void (*ScatterTaskCountdownErrorCallback)(GError *error, gpointer user_data);
+
+typedef struct {
+  guint remaining;
+  ScatterTaskCountdownCompleteCallback callback;
+  ScatterTaskCountdownErrorCallback error_callback;
+  gpointer data;
+} ScatterTaskCountdown;
+
+static ScatterTaskCountdown *
+scatter_task_countdown_new (guint remaining,
+                            ScatterTaskCountdownCompleteCallback callback,
+                            ScatterTaskCountdownErrorCallback error_callback,
+                            gpointer data)
+{
+  ScatterTaskCountdown *countdown = g_new0(ScatterTaskCountdown, 1);
+  countdown->remaining = remaining;
+  countdown->callback = callback;
+  countdown->error_callback = error_callback;
+  countdown->data = data;
+
+  return countdown;
+}
+
+static void
+scatter_task_countdown_subtask_completed (ScatterTaskCountdown **countdown_ptr,
+                                          GError *error)
+{
+  ScatterTaskCountdown *countdown = *countdown_ptr;
+
+  /* If we received an error we should still continue to countdown, but
+   * allow the creator of the countdown to handle the error */
+  if (error && countdown->error_callback)
+    {
+      (*countdown->error_callback)(error, countdown->data);
+    }
+
+  if (--countdown->remaining == 0)
+    {
+      (*countdown->callback)(countdown->data);
+      g_free (countdown);
+      *countdown_ptr = NULL;
+    }
+}
+
+typedef struct {
+  /* Note that scatter_task_countdown_subtask_completed will free
+   * this member once the countdown has completed */
+  ScatterTaskCountdown *countdown;
+  EkncDomain *domain;
+  GPtrArray *models;
+  guint index;
+} ObjectResponseData;
+
+static ObjectResponseData *
+object_response_data_new (ScatterTaskCountdown *countdown,
+                          EkncDomain *domain,
+                          GPtrArray *models,
+                          guint index)
+{
+  ObjectResponseData *data = g_new0(ObjectResponseData, 1);
+  data->countdown = countdown;
+  data->domain = domain;
+  data->models = models;
+  data->index = index;
+  return data;
+}
+
+static void
+object_response_data_free (ObjectResponseData *data)
+{
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC(ObjectResponseData, object_response_data_free);
```

Comment:
On the other hand, since the cleanup function is just `g_free()` you could just use `g_autofree` instead of the autoptr stuff?